[aten] Speed up vec256::loadu

bertmaher opened the pull request:
Summary: Since we know we are loading fewer than 8 elements we can use vmaskmov, which is faster than memset.

Test Plan:
Microbenchmark on a skylake server shows an improvement from 17->10 ns.

Old and busted:
```
---------------------------------------------------------
Benchmark                  Time           CPU Iterations
---------------------------------------------------------
Loadu                     17 ns         17 ns   40529728
```

New hotness:
```
---------------------------------------------------------
Benchmark                  Time           CPU Iterations
---------------------------------------------------------
Loadu                     10 ns         10 ns   68226581
```

Differential Revision: D29699470

colesbury commented on aten/src/ATen/cpu/vec/vec256/vec256_float.h:
```diff
@@ -74,18 +74,26 @@ template <> class Vectorized<float> {
+        _mm256_load_si256(reinterpret_cast<const __m256i*>(masks[count]));
+    return _mm256_maskload_ps(reinterpret_cast<const float*>(ptr), mask);
   }
   void store(void* ptr, int64_t count = size()) const {
```
Do you plan to use the same strategy for `store`?

bertmaher replied:
Good call, I should go ahead and do `store` while I'm here.

colesbury reviewed (commented):
Nice!

bertmaher committed:
[aten] Speed up vec256::loadu (#61667)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/61667

Since we know we are loading fewer than 8 elements we can use vmaskmov, which is faster than memset.

Test Plan:
Microbenchmark on a skylake server shows an improvement from 18->3 ns.

Memcpy
```
--------------------------------------------------
Benchmark           Time           CPU Iterations
--------------------------------------------------
Loadu              18 ns         18 ns   39637568
```

Subtract+compare (less optimal masking approach)
```
--------------------------------------------------
Benchmark           Time           CPU Iterations
--------------------------------------------------
Loadu               5 ns          5 ns  138692501
```

Lookup table (this masking approach)
```
--------------------------------------------------
Benchmark           Time           CPU Iterations
--------------------------------------------------
Loadu               3 ns          3 ns  224088764
```

Differential Revision: D29699470

fbshipit-source-id: 748f3413bc90ef1ba2aa79eb88ca3c966d526745

bertmaher commented:
abandoning this, b/c there are possible memory safety issues with masked loads on some platforms and it's not worth trying to test this absolutely everywhere.

colesbury commented:
What are the memory safety issues? On which platofrms?

ngimel commented:
See comments on the diff

bertmaher commented:
Sorry @colesbury I should've put this finding on the the public PR.  The AMD manual leaves open the possibility that masked-out addresses can still generate page faults:

I usually just read the Intel docs for fault behavior, which is reassuring:
> Faults occur only due to mask-bit required memory accesses that caused the faults. Faults will not occur due to referencing any memory location if the corresponding mask bit for that memory location is 0. For example, no faults will be detected if the mask bits are all zero.

but the AMD docs crush my hopes:
> Exception and trap behavior for elements not selected for loading or storing from/to memory is implementation dependent. For instance, a given implementation may signal a data breakpoint or a page fault for doublewords that are zero-masked and not actually written.
So, damn.  Seems like this could cause a spurious page fault/SEGV if it's used unaligned at the boundary of an unmapped page 

I am sort of skeptical whether it would actually be a problem in practice but it also feels like a bit of a risk for a very small micro-optimization
